crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The paginator is a stateful helper object that handles reaching an event,
+//! either from a cache or network, and surrounding events ("context"). Then, it
+//! makes it possible to paginate forward or backward, from that event, until
+//! one end of the timeline (front or back) is reached.
+
+use eyeball::{SharedObservable, Subscriber};
+use matrix_sdk_base::{deserialized_responses::TimelineEvent, SendOutsideWasm, SyncOutsideWasm};
+use ruma::{api::Direction, uint, EventId, OwnedEventId};
+use tokio::sync::Mutex;
+
+use crate::{
+    room::{EventWithContextResponse, Messages, MessagesOptions},
+    Room,
+};
+
+/// Current state of a [`Paginator`].
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum PaginatorState {
+    /// The initial state of the paginator.
+    Initial,
+
+    /// The paginator is fetching the target initial event.
+    FetchingTargetEvent,
+
+    /// The target initial event could be found, zero or more paginations have
+    /// happened since then, and the paginator is at rest now.
+    Idle,
+
+    /// The paginator is… paginating one direction or another.
+    Paginating,
+}
+
+/// An error that happened when using a [`Paginator`].
+#[derive(Debug, thiserror::Error)]
+pub enum PaginatorError {
+    /// The target event could not be found.
+    #[error("target event with id {0} could not be found")]
+    EventNotFound(OwnedEventId),
+
+    /// We're trying to manipulate the paginator in the wrong state.
+    #[error("expected paginator state {expected:?}, observed {actual:?}")]
+    InvalidPreviousState {
+        /// The state we were expecting to see.
+        expected: PaginatorState,
+        /// The actual state when doing the check.
+        actual: PaginatorState,
+    },
+
+    /// There was another SDK error while paginating.
+    #[error("an error happened while paginating")]
+    SdkError(#[source] crate::Error),
+}
+
+/// A stateful object to reach to an event, and then paginate backward and
+/// forward from it.
+///
+/// See also the module-level documentation.
+pub struct Paginator {
+    /// The room in which we're going to run the pagination.
+    room: Box<dyn PaginableRoom>,
```

Comment:
Curious why you went for `Box` instead of making this struct generic.

Reply:
Force of habit, and it's not performance critical, so the indirect call seemed fine.